src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -646,43 +651,44 @@ private async Task ProcessRequests<TContext>(IHttpApplication<TContext> applicat
 
         public void OnStarting(Func<object, Task> callback, object state)
         {
-            lock (_onStartingSync)
+            if (HasResponseStarted)
             {
-                if (HasResponseStarted)
-                {
-                    ThrowResponseAlreadyStartedException(nameof(OnStarting));
-                }
+                ThrowResponseAlreadyStartedException(nameof(OnStarting));
+            }
 
-                if (_onStarting == null)
-                {
-                    _onStarting = new Stack<KeyValuePair<Func<object, Task>, object>>();
-                }
-                _onStarting.Push(new KeyValuePair<Func<object, Task>, object>(callback, state));
+            if (_onStartingRunning)
```

Comment:
You don't need to add these bools. If you just iterate until TryPop returns false in the FireOnStarting/Completed methods like @pakrym suggested, the newly added callback would get called next. This feels better than running the callback inline in OnStarting/Completed anyway.

Reply:
Honestly, I don't care about this scenario. We just need to decide on the behavior because this isn't the point of this change and I've wasted too much time thinking about it.

Reply:
I agree this scenario isn't too important. So KISS. Removing _onStartingRunning and _onCompletedRunning and using TryPop instead simplifies the code greatly.

Reply:
Done